proptest/src/num/float_samplers.rs

```diff
@@ -0,0 +1,318 @@
+//-
+// Copyright 2022 Michael Wright
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Alternative uniform float samplers because the ones provided by the rand crate are prone
+//! to overflow. The samplers work by uniformly selecting from a set of equally spaced values in
+//! the interval and the included bounds. Selection is slighly biased towards the bounds.
+
+pub(crate) use self::f32::F32U;
+pub(crate) use self::f64::F64U;
+
+macro_rules! float_sampler {
+    ($typ: ident, $int_typ: ident, $wrapper: ident) => {
+        pub mod $typ {
+            use rand::prelude::*;
+            use rand::distributions::uniform::{
+                SampleBorrow, SampleUniform, Uniform, UniformSampler,
+            };
+
+            #[must_use]
+            // Returns the previous float value. In other words the greatest value representable
+            // as a float such that `last_float(a) < a`. `-0.` is treated as `0.`.
+            fn last_float(a: $typ) -> $typ {
+                debug_assert!(a.is_finite() && a > $typ::MIN, "`last_float` invalid input: {}", a);
+                if a == (0.) {
+                    -$typ::from_bits(1)
+                } else if a < 0. {
+                    $typ::from_bits(a.to_bits() + 1)
+                } else {
+                    $typ::from_bits(a.to_bits() - 1)
+                }
+            }
+
+            #[must_use]
+            fn last_gap(a: $typ) -> $typ {
+                debug_assert!(a.is_finite() && a > $typ::MIN, "`last_float` invalid input: {}", a);
+                a - last_float(a)
+            }
+
+            #[derive(Copy, Clone, Debug)]
+            pub(crate) struct $wrapper($typ);
+
+            impl From<$typ> for $wrapper {
+                fn from(x: $typ) -> Self {
+                    $wrapper(x)
+                }
+            }
+            impl From<$wrapper> for $typ {
+                fn from(x: $wrapper) -> Self {
+                    x.0
+                }
+            }
+
+            #[derive(Clone, Copy, Debug)]
+            pub(crate) struct FloatUniform {
+                uniform: Uniform<$int_typ>,
+                values: SampleValueCollection,
+            }
+
+            impl UniformSampler for FloatUniform {
+
+                type X = $wrapper;
+
+                fn new<B1, B2>(low: B1, high: B2) -> Self
+                where
+                    B1: SampleBorrow<Self::X> + Sized,
+                    B2: SampleBorrow<Self::X> + Sized,
+                {
+                    let low = low.borrow().0;
+                    let high = high.borrow().0;
+
+                    let values = SampleValueCollection::new(low, last_float(high));
```

Comment:
I'm pretty sure theres a minor bug here since we pass in `last_float(high)` here, and then `SampleValueCollection::new` also calculates `last_float(high)` with whats passed in so we actually end up using `last_float(last_float(high))`. Proposing we remove `SampleValueCollection::new`, only have `SampleValueCollection::new_inclusive`, and keep passing in `last_float(high)`here